src/libraries/Common/src/System/Net/Sockets/SocketConnection.cs

```diff
@@ -0,0 +1,124 @@
+// Licensed to the .NET Foundation under one or more agreements.
+// The .NET Foundation licenses this file to you under the MIT license.
+
+using System.Diagnostics.CodeAnalysis;
+using System.IO;
+using System.IO.Pipelines;
+using System.Net.Sockets;
+using System.Runtime.ExceptionServices;
+using System.Threading;
+using System.Threading.Tasks;
+
+namespace System.Net.Connections
+{
+    internal sealed class SocketConnection : Connection, IConnectionProperties
+    {
+        private readonly Socket _socket;
+        private readonly IDataChannelProvider _factory;
+        private Stream? _stream;
+        private IDuplexPipe? _pipe;
+        private readonly IConnectionProperties? _options;
+
+        public override EndPoint? RemoteEndPoint => _socket.RemoteEndPoint;
+        public override EndPoint? LocalEndPoint => _socket.LocalEndPoint;
+        public override IConnectionProperties ConnectionProperties => this;
+
+        public SocketConnection(Socket socket, IDataChannelProvider factory, IConnectionProperties? options)
```

Comment:
Consider using a partial class instead of `IDataChannelProvider` to avoid interface dispatch?